onnxruntime/core/providers/webgpu/tensor/depth_to_space.cc

```diff
@@ -0,0 +1,145 @@
+
+  return Status::OK();
+}
+
+template <bool is_nhwc>
+Status DepthToSpace<is_nhwc>::ComputeInternal(onnxruntime::webgpu::ComputeContext& context) const {
+  const auto* input = context.Input(0);
+  const TensorShape input_shape = input->Shape();
+
+  int64_t n, c, h, w;
+  int64_t shape[6];
+  int64_t perm[6];
+  if (is_nhwc) {
+    n = input_shape[0];
+    h = input_shape[1];
+    w = input_shape[2];
+    c = input_shape[3];
+
+    if (is_dcr_) {
+      int64_t shape_values[] = {n, h, w, blocksize_, blocksize_, c / (blocksize_ * blocksize_)};
+      int64_t perm_values[] = {0, 1, 3, 2, 4, 5};
+      std::copy(shape_values, shape_values + 6, shape);
+      std::copy(perm_values, perm_values + 6, perm);
+    } else {
+      int64_t shape_values[] = {n, h, w, c / (blocksize_ * blocksize_), blocksize_, blocksize_};
+      int64_t perm_values[] = {0, 1, 4, 2, 5, 3};
+      std::copy(shape_values, shape_values + 6, shape);
+      std::copy(perm_values, perm_values + 6, perm);
+    }
+  } else {
+    n = input_shape[0];
+    h = input_shape[2];
+    w = input_shape[3];
+    c = input_shape[1];
+
+    if (is_dcr_) {
+      int64_t shape_values[] = {n, blocksize_, blocksize_, c / (blocksize_ * blocksize_), h, w};
+      int64_t perm_values[] = {0, 3, 4, 1, 5, 2};
+      std::copy(shape_values, shape_values + 6, shape);
+      std::copy(perm_values, perm_values + 6, perm);
+    } else {
+      int64_t shape_values[] = {n, c / (blocksize_ * blocksize_), blocksize_, blocksize_, h, w};
+      int64_t perm_values[] = {0, 1, 4, 2, 5, 3};
+      std::copy(shape_values, shape_values + 6, shape);
+      std::copy(perm_values, perm_values + 6, perm);
+    }
+  }
+
+  std::vector<int64_t> shape_vec(shape, shape + 6);
+  TensorShape input_override_shape(shape_vec);
+
+  // Calculate the final 4D output shape
+  int64_t output_shape[4];
+  if (is_nhwc) {
+    int64_t output_shape_values[] = {n, h * blocksize_, w * blocksize_, c / (blocksize_ * blocksize_)};
+    std::copy(output_shape_values, output_shape_values + 4, output_shape);
+  } else {
+    int64_t output_shape_values[] = {n, c / (blocksize_ * blocksize_), h * blocksize_, w * blocksize_};
+    std::copy(output_shape_values, output_shape_values + 4, output_shape);
+  }
+  TensorShape final_output_shape(gsl::make_span(output_shape));
+
+  auto* output = context.Output(0, final_output_shape);
+  int64_t output_size = output->Shape().Size();
+
+  if (output_size == 0) {
+    return Status::OK();
+  }
+
+  std::vector<int64_t> shape_after_permutation_vec(6);
+  for (int i = 0; i < 6; i++) {
+    shape_after_permutation_vec[i] = shape[perm[i]];
+  }
+  TensorShape output_override_shape(shape_after_permutation_vec);
+
+  DepthToSpaceProgram program{perm};
+  program
+      .AddInput({input, ProgramTensorMetadataDependency::TypeAndRank, input_override_shape, 1})
+      .AddOutput({output, ProgramTensorMetadataDependency::TypeAndRank, output_override_shape, 1})
```

Comment:
since input and output are of the same type `T`, you need either input or output depending on type should be ok.

also, it looks like input and output MUST be rank==4, so maybe remove rank here too. (what will happen if input is not rank-4?)